### PR TITLE
International Workshop Attendance Form: More reliable workshop date validation

### PR DESCRIPTION
**Pull request by islemaster**
Update to the [International Workshop Attendance Form](https://studio.code.org/pd/international_workshop). Right now the workshop date ('date' on the model) looks required in the client UI but the form can be submitted when the textbox is blank, and a date (today) is still sent to the server. We're also not requiring the date on the model so blank submissions are possible.

Two changes:
1. Require this field on the model (and require that it is a parse-able date)
2. Make sure the user experience on the client is consistent with the field being "Required"

Fixes [PLC-91](https://codedotorg.atlassian.net/browse/PLC-91).

I'm only halfway happy with the client changes here, but I think the behavior is better than what we had before.  The date picker reflects the underlying state we're about to submit _most_ of the time. That underlying state now gets cleared to `null` if we're not able to parse a date out of the input, which results in a validation error on submission (since the field is properly required now).

peek 2019-02-20 11-03

Styling after submitting blank/invalid input:
screenshot from 2019-02-20 11-01-46

There's still an odd edge case where you can delete part of the date and `moment` will try to fill in the rest with a sensible default.  If, when this happens, the underlying date is not different from the previous value, the value in the text field doesn't update to show a whole date but the hidden state (and submitted value) is still the previous value.

peek 2019-02-20 11-13

I'm pretty sure this is a bug in the [react-datepicker](https://reactdatepicker.com) component, and has been fixed in more recent versions.  In fact, there's a generally broken paradigm here where the text field is "half-controlled" in that you hook it up like a controlled/stateless component using `Date` values, but you're not guaranteed that the textbox will always reflect a valid date after a blur event. The latest version (2.1.0) of this component doesn't have this problem - it seems to provide a stronger guarantee that the component will always be left either blank or representing a complete, valid date when not being actively editing.  We should definitely upgrade one of these days.

### CODE REVIEW

**Comment by islemaster** [comment]
Asking Brendan and Elijah for review since they've worked on this code, but FYI @clareconstantine and @hacodeorg since this is in our space.

File: apps/src/code-studio/pd/workshop_dashboard/components/date_picker.jsx [hunk 1]
```diff
@@ -95,6 +95,7 @@ export default class DatePicker extends React.Component {
   static propTypes = {
     date: PropTypes.object,
     onChange: PropTypes.func.isRequired,
+    onBlur: ReactDatePicker.propTypes.onBlur,
```
Comment (Hamms) [inline apps/src/code-studio/pd/workshop_dashboard/components/date_picker.jsx:98]: huh, this is a cool pattern. Should we update the other props that are passed directly through to ReactDatePicker to mirror this?

File: dashboard/test/models/pd/international_opt_in_test.rb [hunk 2]
```diff
@@ -35,4 +36,23 @@ class Pd::InternationalOptInTest < ActiveSupport::TestCase
...
+    malformed_date = build :pd_international_opt_in,
+      user_id: teacher.id,
+      form_data: FORM_DATA.merge({date: 'malformed-date'}).to_json
+    refute malformed_date.valid?
+
+    valid_date = build :pd_international_opt_in,
+      user_id: teacher.id, form_data:
+        FORM_DATA.merge({date: '2019-02-18'}).to_json
+    assert valid_date.valid?
+  end
```
Comment (Hamms) [inline dashboard/test/models/pd/international_opt_in_test.rb:57]: :+1:

File: apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx [hunk 3]
```diff
@@ -55,18 +55,30 @@ class InternationalOptInComponent extends FormComponent {
...
+  };
+
+  workshopDateAsMoment() {
+    const {data} = this.props;
+    const date = data && moment(data.date, DATE_FORMAT);
+    if (date && date.isValid()) {
+      return date;
+    }
+    return null;
+  }
```
Comment (Hamms) [inline apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: could this logic be combined with handleDateChange? maybe something like

```js
validateDate(date) {
  date = moment(date, DATE_FORMAT);
  if (date && date.isValid()) {
    return date;
  }
  return null;
}

handleDateChange = date => {
  super.handleChange({ date: this.validateDate(date) });
}

...

<DatePicker
  date={this.validateDate(this.props.data && this.props.data.date)}
```

**Review by Hamms — APPROVED** [review]
couple minor suggestions, but generally LGTM!

**Review by breville — APPROVED** [review]

File: apps/src/code-studio/pd/workshop_dashboard/components/date_picker.jsx (hunk 1 above)
Comment (islemaster) [reply apps/src/code-studio/pd/workshop_dashboard/components/date_picker.jsx:98]: :+1: I'd be in favor of that!  Or even something more ridiculous, calling out that we're just passing through loads of these:
```jsx
static propTypes = {
  date: PropTypes.object,
  onChange: PropTypes.func.isRequired,
  ..._.pick(ReactDatePicker.propTypes, [
    'onBlur',
    'minDate',
    'maxDate',
    'selectsStart',
    // ...
  ]),
};
```

File: apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx (hunk 3 above)
Comment (islemaster) [reply apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx]: Hm - not quite.  I have to pass a date string to `handleChange` but the DatePicker component wants a `Moment` object.  What I've done instead is extract a pure helper like you've suggested here, it does seem cleaner.

**Pull request merged**